lib/Corres_Cases.thy

```diff
@@ -0,0 +1,319 @@
+   These lemmas bind the corres guards to the corres_case_helper guards. *)
+private lemma corres_case_helper_corres_left:
+  "corres_underlying sr nf nf' r Q Q' f f' \<Longrightarrow>
+    corres_case_helper (P, P') (Q, Q') (corres_underlying sr nf nf' r P P' f f')"
+  by (auto simp: corres_case_helper_def elim!: corres_guard_imp)
+
+private lemma corres_case_helper_corres_right:
+  "corres_underlying sr nf nf' r Q' Q f f' \<Longrightarrow>
+    corres_case_helper (P, P') (Q, Q') (corres_underlying sr nf nf' r P' P f f')"
+  by (auto simp: corres_case_helper_def elim!: corres_guard_imp)
+
+
+section \<open>Main method definitions\<close>
+
+(* Case distinction on abstract side *)
+method corres_cases_left =
+  determ \<open>
+    corres_pre,
+    (match conclusion in
+      "corres_underlying sr nf nf' r P P' (f >>= g) f'" for sr nf nf' r P P' f g f'
+         \<Rightarrow> \<open>apply_split f "\<lambda>f. corres_underlying sr nf nf' r P P' (f >>= g) f'"\<close>
+     \<bar> "corres_underlying sr nf nf' r P P' (f >>=E g) f'" for sr nf nf' r P P' f g f'
+         \<Rightarrow> \<open>apply_split f "\<lambda>f. corres_underlying sr nf nf' r P P' (f >>=E g) f'"\<close>
+     \<bar>  "corres_underlying sr nf nf' r P P' f f'" for sr nf nf' r P P' f f'
+         \<Rightarrow> \<open>apply_split f "\<lambda>f. corres_underlying sr nf nf' r P P' f f'"\<close>),
+    corres_cases_body helper: corres_case_helper_corres_left\<close>
+
+(* case distinction on concrete side *)
+method corres_cases_right =
+  determ \<open>
+    corres_pre,
+    (match conclusion in
+      "corres_underlying sr nf nf' r P P' f (f' >>= g)" for sr nf nf' r P P' f g f'
+         \<Rightarrow> \<open>apply_split f' "\<lambda>f'. corres_underlying sr nf nf' r P P' f (f' >>= g)"\<close>
+     \<bar> "corres_underlying sr nf nf' r P P' f (f' >>=E g)" for sr nf nf' r P P' f g f'
+         \<Rightarrow> \<open>apply_split f' "\<lambda>f'. corres_underlying sr nf nf' r P P' f (f' >>=E g)"\<close>
+     \<bar>  "corres_underlying sr nf nf' r P P' f f'" for sr nf nf' r P P' f f'
+         \<Rightarrow> \<open>apply_split f' "\<lambda>f'. corres_underlying sr nf nf' r P P' f f'"\<close>),
+     corres_cases_body helper: corres_case_helper_corres_right\<close>
+
+(* single case distinction on either left or right, whichever works first *)
+method corres_cases = corres_cases_left | corres_cases_right
+
+(* Case distinction on abstract and concrete side with quadractic blowup, but attempt to solve
+   contradictory side conditions by simp. Cases that are solved by simp will produce \<top> as guard
+   so that no free schematics are introduced into later goals. *)
+method corres_cases_both uses simp =
+  (* corres_pre first, so that the ";" later only refers to corres goals, not the final implications *)
+  determ \<open>
+    corres_pre,
+    (corres_cases_left; corres_cases_right;
+     (solves \<open>rule corres_inst[where P=\<top> and P'=\<top>], simp add: simp\<close>)?)\<close>
```

Comment:
This one is really interesting, and I think very clever.

Reply:
This is actually even a tiny bit more powerful than I thought it might be, because we have global simp rules that will solve trivial `corres` statements like `corres .. (return x) (return x)`, so these might also disappear, but only if they are solvable with a trivial guard. I only found that out later in the VSpace examples.

Reply:
It's not for this PR, but it might be worth auditing those simp rules at some point and thinking about whether they should have schematic or fixed preconditions.

It doesn't normally break proofs but I've had instances before where I was annoyed about schematics propagating further than expected due to a trivial `corres` statement not instantiating its preconditions to `⊤`.

Reply:
Yes, it's super annoying, and in the upcoming `corres` method they definitely cause trouble (Dan encountered the same and has a set of "simp rules that break everything"). I was too afraid of a big update to try to get rid of them properly, because I have the feeling they are used almost everywhere where we have a final `return` or `gets`.

Might be best to schedule that update for when we have arch split Refine, then we at least will have to do only one pass and not five.